File: Classes/Core Data Categories/ScheduledShow+TV.m
//
//  ScheduledShow+TV.m
//  Signal GH
//
//  Created by Glenn Howes on 2/4/14.
//  Copyright (c) 2014 Generally Helpful Software. All rights reserved.
//

#import "ScheduledShow+TV.h"
#import "EventInformationTable.h"
#import "ExtendedTextTable.h"
#import "SystemTimeTable.h"
#import "LocalizedString+TV.h"
#import "ShowTitle.h"
#import "ShowDescription.h"
#import "TunerSubchannel+TV.h"
#import "TunerChannel+TV.h"
#import "LanguageString.h"
#import "ContentAdvisoryDescriptor.h"
#import "ContentAdvisory+TV.h"

@implementation NSDate(DigitalTV)


+(UInt64) startTimeNearestMinute:(NSDate*)baseTime
{
    NSTimeInterval seconds = [baseTime timeIntervalSince1970];
    seconds = rint(seconds);
    UInt64 result = seconds+30;
    result /= 60;
    result *= 60; // round to minute
    return result;
}


-(NSDate*) dateAtNearestMinute
{
    NSTimeInterval timeIntervalNearestMinute = [NSDate startTimeNearestMinute:self];
    NSDate* result = self;
    if([result timeIntervalSince1970] != timeIntervalNearestMinute)
    {
        result = [[NSDate alloc] initWithTimeIntervalSince1970:timeIntervalNearestMinute];
    }

    return result;
}
@end


@implementation ScheduledShow (TV)

-(NSString*)title
{
    NSSet* myTitles = self.titles;
    
    NSString* result = [LocalizedString bestMatchFromSet:myTitles];
    return result;
}

-(NSString*) advisoryString
{
    NSString* result = nil;
    ContentAdvisory* relevantContentAdvisory = [ContentAdvisory retrieveBestMatchFromArrayOfAdvisories:[self.contentAdvisories allObjects]];
    if(relevantContentAdvisory != nil)
    {
        result = [relevantContentAdvisory advisoryStringGivenSetOfRatings:self.subChannel.channel.ratings];
    }
    
    return result;
}

-(NSString*) showDescription
{
    NSSet* myTitles = self.descriptions;
    
    NSString* result = [LocalizedString bestMatchFromSet:myTitles];
    return result;
}


-(NSDateFormatter*) timeFormatter  // main thread only
{
    static NSDateFormatter *sResult = nil;
    
    if(sResult == nil)
    {
        sResult = [[NSDateFormatter alloc] init];
        
        [sResult setTimeStyle:NSDateFormatterShortStyle];
        [sResult setDateStyle:NSDateFormatterMediumStyle];
        
        [sResult setDoesRelativeDateFormatting:YES];
    }
    return sResult;
}

-(NSString*) startTimeString
{
    NSString* result = nil;
    NSDate* now = [NSDate date];
    NSDate* startTime = self.start_time;
    
    
    NSTimeInterval seconds = [startTime timeIntervalSinceNow];
    NSTimeInterval minutes = rint(seconds/60);
    NSInteger intMinutes = minutes;
    
    if([startTime compare:now] == NSOrderedDescending )
    {
        if(intMinutes == 0)
        {
            result = NSLocalizedString(@"Starting Now.", @"");
        }
        else if(intMinutes == 1)
        {
            result = NSLocalizedString(@"Starts in a minutes.", @"");
        }
        else if(intMinutes < 60)
        {
            NSString* formatString = NSLocalizedString(@"Starts in %d minutes.", @"");
            result = [NSString stringWithFormat:formatString, intMinutes];
        }
        else
        {
            NSString* endDateString = [[self timeFormatter] stringFromDate:startTime];
            
            NSString* formatString = NSLocalizedString(@"Starts %@.", @"");
            result = [NSString stringWithFormat:formatString, endDateString];
        }
    }
    else if([startTime compare:now] == NSOrderedAscending)
    {
        int deltaMinutes = abs((int)intMinutes);
        if(deltaMinutes == 0)
        {
            result = NSLocalizedString(@"Just Started.", @"");
        }
        else if(deltaMinutes == 1)
        {
            result = NSLocalizedString(@"Started a minute ago.", @"");
        }
        else if(deltaMinutes < 60)
        {
            NSString* formatString = NSLocalizedString(@"Started %d minutes ago.", @"");
            result = [NSString stringWithFormat:formatString, deltaMinutes];
        }
        else
        {
            NSString* endDateString = [[self timeFormatter] stringFromDate:startTime];
            
            NSString* formatString = NSLocalizedString(@"Started %@.", @"");
            result = [NSString stringWithFormat:formatString, endDateString];
        }
    }
    else
    {
        result = NSLocalizedString(@"Starting Now", @"");
    }
    
    return result;
}

-(NSString*) endTimeString
{
    NSString* result = nil;
    NSDate* now = [NSDate date];
    NSDate* startTime = self.start_time;
    NSDate* endTime = self.end_time;
    
    if([startTime compare:now] != NSOrderedDescending && [endTime compare:now] != NSOrderedAscending)
    {
        NSTimeInterval seconds = [endTime timeIntervalSinceNow];
        NSTimeInterval minutes = rint(seconds/60);
        NSInteger intMinutes = minutes;
        if(intMinutes == 1)
        {
            result = NSLocalizedString(@"Ends this minute", @"");
        }
        else if(intMinutes == 0)
        {
            if(endTime == nil)
            {
                result = @"";
            }
            else
            {
                result = NSLocalizedString(@"Ends Now", @"");
            }
        }
        else
        {
            NSString* formatString = NSLocalizedString(@"Ends in %d minutes.", @"");
            result = [NSString stringWithFormat:formatString, intMinutes];
        }
    }
    else if([endTime compare:now] == NSOrderedAscending)
    {
        NSString* endDateString = [[self timeFormatter] stringFromDate:endTime];
        
        NSString* formatString = NSLocalizedString(@"Ended %@.", @"");
        result = [NSString stringWithFormat:formatString, endDateString];
    }
    else
    {
        NSString* endDateString = [[self timeFormatter] stringFromDate:endTime];
        
        NSString* formatString = NSLocalizedString(@"Ends %@.", @"");
        result = [NSString stringWithFormat:formatString, endDateString];
    }
    
    return result;
}


+(NSDate*) startDateFromEventRecord:(EventInformationRecord*)eventRecord withTimeOffest:(NSTimeInterval)timeOffset
{
    NSTimeInterval startTime = eventRecord.start_time;
    
    NSTimeInterval dateTime = startTime+timeOffset+[SystemTimeTable beginningOf1980];
    NSDate* result = [[NSDate alloc] initWithTimeIntervalSince1970:dateTime];
    result = [result dateAtNearestMinute];
    return result;
}

+(NSDate*) endDateFromEventRecord:(EventInformationRecord*)eventRecord withTimeOffest:(NSTimeInterval)timeOffset
{
    NSDate* start_time = [ScheduledShow startDateFromEventRecord:eventRecord withTimeOffest:timeOffset];
    
    NSDate* result = [start_time dateByAddingTimeInterval:eventRecord.length_in_seconds];
    return result;
}


-(void) updateFromEventRecord:(EventInformationRecord*)eventRecord withTimeOffest:(NSTimeInterval)timeOffset;
{
    NSTimeInterval startTime = eventRecord.start_time;
    
    NSTimeInterval dateTime = startTime+timeOffset+[SystemTimeTable beginningOf1980];
    NSDate* start_time = [[NSDate alloc] initWithTimeIntervalSince1970:dateTime];
    start_time = [start_time dateAtNearestMinute];
    if(self.start_time == nil || fabs([self.start_time timeIntervalSinceDate:start_time]) > 120)
    {
           self.start_time = start_time;
    }
    NSDate* end_time = [start_time dateByAddingTimeInterval:eventRecord.length_in_seconds];
    self.end_time = end_time;
    self.event_id = [NSNumber numberWithInteger:eventRecord.event_id];
    
    NSEntityDescription *showTitleEntity = [NSEntityDescription entityForName:@"ShowTitle" inManagedObjectContext:[self managedObjectContext]];
    
    for(LanguageString* aLanguageString in eventRecord.titles)
    {
        BOOL foundIt = NO;
        for(ShowTitle* aTitle in self.titles)
        {
            if([aTitle.locale isEqualToString:aLanguageString.languageCode])
            {
                foundIt = YES;
                aTitle.text = aLanguageString.string;
                break;
            }
        }
        if(!foundIt)
        {
            ShowTitle* newTitle = [[ShowTitle alloc] initWithEntity:showTitleEntity insertIntoManagedObjectContext:self.managedObjectContext];
            newTitle.locale = aLanguageString.languageCode;
            newTitle.text = aLanguageString.string;
            newTitle.show = self;
            [self addTitlesObject:newTitle];
        }
    }
    if(self.contentAdvisories.count == 0)
    {
        for(ATSCTable* aTable in eventRecord.descriptors)
        {
            if([aTable isKindOfClass:[ContentAdvisoryDescriptor class]])
            {
                ContentAdvisoryDescriptor* contentAdvisoryTable = (ContentAdvisoryDescriptor*)aTable;
                [ContentAdvisory extractAdvisoriesFromDescriptors:contentAdvisoryTable intoShow:self];
            }
        }
    }
}

-(void) updateFromExtendedTextTable:(ExtendedTextTable*)extendedText
{
    NSEntityDescription *showDescriptionEntity = [NSEntityDescription entityForName:@"ShowDescription" inManagedObjectContext:[self managedObjectContext]];
    for(LanguageString* aLanguageString in extendedText.strings)
    {
        BOOL foundIt = NO;
        for(ShowDescription* aDescription in self.descriptions)
        {
            if([aDescription.locale isEqualToString:aLanguageString.languageCode] && aLanguageString.string.length)
            {
                foundIt = YES;
                aDescription.text = aLanguageString.string;
                break;
            }
        }
        if(!foundIt)
        {
            ShowDescription* newDescription = [[ShowDescription alloc] initWithEntity:showDescriptionEntity insertIntoManagedObjectContext:self.managedObjectContext];
            newDescription.locale = aLanguageString.languageCode;
            newDescription.text = aLanguageString.string;
            newDescription.show = self;
            [self addDescriptionsObject:newDescription];
        }
    }
}


-(CGFloat)descriptionHeightForWidth:(CGFloat)width
{
    CGFloat result = 0.0;
    NSString* description = self.showDescription;
    if(description.length)
    {
        NSStringDrawingContext* drawingContext = [[NSStringDrawingContext alloc] init];
        UIFontDescriptor* fontDescriptor = [UIFontDescriptor preferredFontDescriptorWithTextStyle:UIFontTextStyleCaption1];
        CGSize fitToSize = CGSizeMake(width, 480);
        CGRect boundingRect = [description boundingRectWithSize:fitToSize options:NSStringDrawingTruncatesLastVisibleLine | NSStringDrawingUsesLineFragmentOrigin attributes:fontDescriptor.fontAttributes context:drawingContext];
        result = boundingRect.size.height;
    }
    return result;
}

@end

File: Classes/Core Data Categories/ScheduledShow+TV.swift
//
//  ScheduledShow+TV.swift
//  Signal GH
//
//  Created by Glenn Howes on 4/8/17.
//  Copyright © 2017 Generally Helpful Software. All rights reserved.
//

import Foundation
import CoreData
import CoreGraphics
import SVGgh

public extension ScheduledShow
{
    @objc public var networkIcon : SVGRenderer?
    {
        let result = self.subChannel?.networkIcon
        return result
    }
    

    
    @objc public var twitterText : String
    {
        var result = ""
        self.managedObjectContext?.performAndWait
        {
            let newLineString = "\n"
            var strings = Array<String>()
            if let title = self.title
            {
                strings.append(title)
            }
            
            
            if let subchannelDescription = self.subChannel?.completedDescription
            {
                strings.append(subchannelDescription)
            }
            
            if let description = self.showDescription
            {
                strings.append(description)
            }
            
            
            if let endDate = self.end_time, let startDate = self.start_time
            {
                let dateFormatter = DateFormatter()
                dateFormatter.timeStyle = .short
                dateFormatter.dateStyle = .medium
                
                dateFormatter.doesRelativeDateFormatting = true
                
                let timeZone = NSTimeZone.system
                
                
                let starts = NSLocalizedString("Starts: ", comment: "") + dateFormatter.string(from: (startDate as Date) as Date) + " " + timeZone.localizedName(for: timeZone.isDaylightSavingTime() ? .shortDaylightSaving : .shortStandard , locale: NSLocale.system)!
                let length = Int(floor(endDate.timeIntervalSince(startDate as Date)/60.0))
                let lasts = NSLocalizedString("Lasts: ", comment: "") + String(length) + NSLocalizedString(" min", comment: "")
                strings.append(starts)
                strings.append(lasts)

            }
            
            for aString in strings
            {
                let length = aString.count
                let runningLength = result.count
                let addedLength = (runningLength > 0) ? length+1 : length
                if (addedLength + runningLength) > 280
                {
                    continue
                }
                else if runningLength > 0
                {
                    result += newLineString
                    result += aString
                }
                else
                {
                    result = aString
                }
                
            }
            
        }
        return result
    }
    
    @objc public var shareableText : NSAttributedString
    {
        let mutableResult = NSMutableAttributedString()
        
        self.managedObjectContext?.performAndWait {
            let headerFont = UIFont.preferredFont(forTextStyle: .headline)
            let textFont = UIFont.preferredFont(forTextStyle: .body)
            let captionFont = UIFont.preferredFont(forTextStyle: .caption1)
            let captionDescriptor = UIFontDescriptor.preferredFontDescriptor(withTextStyle: .caption1)
            let boldCaptionDescriptor = captionDescriptor.withSymbolicTraits(.traitBold) ?? captionDescriptor
            let boldCaptionFont = UIFont(descriptor: boldCaptionDescriptor, size: 0.0)
            let headerDescription = [NSAttributedStringKey.font : headerFont]
            let textDescription = [NSAttributedStringKey.font : textFont]
            let captionDescription = [NSAttributedStringKey.font :  captionFont]
            let caption2Description = [NSAttributedStringKey.font : boldCaptionFont]
            
            if let title = self.title
            {
                let titleString = NSAttributedString(string: title, attributes: headerDescription)
                mutableResult.append(titleString)
            }
            if let description = self.showDescription
            {
                let addedString = "\n" + description
                let descriptionString = NSAttributedString(string: addedString, attributes: textDescription)
                mutableResult.append((descriptionString))
            }
            
            if let endDate = self.end_time, let startDate = self.start_time
            {
                let dateFormatter = DateFormatter()
                dateFormatter.timeStyle = .short
                dateFormatter.dateStyle = .medium
                
                dateFormatter.doesRelativeDateFormatting = true
                
                let timeZone = NSTimeZone.system
                
                mutableResult.append(NSAttributedString(string: "\n"))
                mutableResult.append(NSAttributedString(string: NSLocalizedString("Starts: ", comment:""), attributes: caption2Description))
                
                let starts = dateFormatter.string(from: (startDate as Date) as Date) + " " + timeZone.localizedName(for: timeZone.isDaylightSavingTime() ? .shortDaylightSaving : .shortStandard , locale: NSLocale.system)!
                mutableResult.append(NSAttributedString(string: starts, attributes: captionDescription))
                
                
                mutableResult.append(NSAttributedString(string: "\n"))
                mutableResult.append(NSAttributedString(string: NSLocalizedString("Lasts: ", comment:""), attributes: caption2Description))
                
                let length = Int(floor(endDate.timeIntervalSince(startDate as Date)/60.0))
                let lasts = String(length) + NSLocalizedString(" min", comment: "")
                mutableResult.append(NSAttributedString(string: lasts, attributes: captionDescription))
            }
        }
        
        return mutableResult.copy() as! NSAttributedString
    }
    
    @objc public var title : String?
    {
        var result : String? = nil
        self.managedObjectContext?.performAndWait {
            guard  let titles = self.titles as?  Set<LocalizedString> else
            {
                return
            }
            result = LocalizedString.bestMatch(fromSet: titles )
        }
        return result
    }
    
    public var advisoryString : String?
    {
        var result : String? = nil
        self.managedObjectContext?.performAndWait {
            guard  let advisories = self.contentAdvisories, let relevantContentAdvisory = ContentAdvisory.retrieveBestMatch(fromAdvisories: advisories as! Set<ContentAdvisory>), let ratings = self.subChannel?.channel?.ratings as? Set<Rating> else
            {
                return
            }
            result = relevantContentAdvisory.advisoryString(givenRatings: ratings)
        }
        return result
    }
    
    public var showDescription : String?
    {
        var result : String? = nil
        self.managedObjectContext?.performAndWait {
            guard  let descriptions = self.descriptions else
            {
                return
            }
            result = LocalizedString.bestMatch(fromSet: descriptions as! Set<LocalizedString>)
        }
        return result
    }
    
    @nonobjc static var timeFormatter : DateFormatter =
    {
        var result = DateFormatter()
        result.timeStyle = .short
        result.dateStyle = .medium
        result.doesRelativeDateFormatting = true
        return result
    }()
    
    public var startTimeString : String?
    {
        var result : String? = nil
        self.managedObjectContext?.performAndWait {
            guard  let startTime = self.start_time else
            {
                return
            }
            let now = Date()
            let seconds = startTime.timeIntervalSinceNow
            let minutes = rint(seconds/60)
            let intMinutes = Int(minutes)
            let deltaMinutes = abs(intMinutes)
            
            switch (startTime.compare(now), minutes)
            {
                case (.orderedDescending, 0), (.orderedSame, _):
                    result = NSLocalizedString("Starting Now", comment: "")
                case (.orderedDescending, 1):
                    result = NSLocalizedString("Starts in a minute", comment: "")
                case (.orderedDescending, 2..<60):
                    let template = NSLocalizedString("Starts in %d minutes", comment: "")
                    
                    result = String(format: template, intMinutes)
                case (.orderedDescending, _):
                    let endDateString = ScheduledShow.timeFormatter.string(from: startTime as Date)
                    let formatString = NSLocalizedString("Starts %@", comment: "")
                    result = String(format: formatString, endDateString)
                case (.orderedAscending, 0):
                    result = NSLocalizedString("Just Started", comment: "")
                case (.orderedAscending, -1):
                    result = NSLocalizedString("Started a minute ago", comment: "")
                case (.orderedAscending, -59..<0):
                    let template = NSLocalizedString("Started %d minutes ago", comment: "")
                    result = String(format: template, deltaMinutes)
                case (.orderedAscending, _):
                    let endDateString = ScheduledShow.timeFormatter.string(from: startTime as Date)
                    let formatString = NSLocalizedString("Started %@", comment: "")
                    result = String(format: formatString, endDateString)
            }
        }
        return result
    }
    
    public func bridges(date testDate: NSDate) -> Bool
    {
        var result = false
        self.managedObjectContext?.performAndWait {
            guard let endTime = self.end_time, let startTime = self.start_time else
            {
                return
            }
            switch (endTime.compare(testDate as Date), startTime.compare(testDate as Date))
            {
                case (.orderedAscending, _), (_, .orderedDescending):
                    result = false
                default:
                    result = true
            }
        }
        return result
    }
    
    public var endTimeString : String?
    {
        var result: String? = nil
        let now = Date()
        let bridgesDate = self.bridges(date: now as NSDate)
        
        self.managedObjectContext?.performAndWait
        {
            guard let endTime = self.end_time else
            {
                return
            }
            
            if bridgesDate
            {
                let seconds = endTime.timeIntervalSinceNow
                let minutes = rint(seconds/60)
                let intMinutes = Int(minutes)
                switch(intMinutes)
                {
                    case 0:
                        result = NSLocalizedString("Ends now", comment: "")
                    case 1:
                        result = NSLocalizedString("Ends this minute", comment: "")
                    default:
                        let template = NSLocalizedString("Ends in %d minutes", comment: "")
                        result = String(format: template, intMinutes)
                }
            }
            else
            {
                let endDateString = ScheduledShow.timeFormatter.string(from: endTime as Date)
                switch endTime.compare(now as Date)
                {
                    case .orderedAscending:
                        let formatString = NSLocalizedString("Ended %@", comment: "")
                        result = String(format: formatString, endDateString)
                    default:
                            let formatString = NSLocalizedString("Ends %@", comment: "")
                    result = String(format: formatString, endDateString)
                    
                }
            }
            
        }
        return result
    }
    
    @nonobjc static func startDate(fromEventRecord eventRecord: EventInformationRecord, withTimeOffset timeOffset: TimeInterval) -> Date
    {
        let startTime = eventRecord.start_time
        let dateTime = startTime+timeOffset + SystemTimeTable.beginningOf1980()
        let baseResult = Date(timeIntervalSince1970: dateTime)
        let result = baseResult.dateAtNearestMinute
        return result
    }
    
    @nonobjc static func endDate(fromEventRecord eventRecord: EventInformationRecord, withTimeOffset timeOffset: TimeInterval) -> Date
    {
        let startTime = ScheduledShow.startDate(fromEventRecord: eventRecord, withTimeOffset: timeOffset)
        let result = startTime.addingTimeInterval(eventRecord.length_in_seconds)
        return result
    }
    
//    +(NSDate*) startDateFromEventRecord:(EventInformationRecord*)eventRecord withTimeOffest:(NSTimeInterval)timeOffset
//    {
//    NSTimeInterval startTime = eventRecord.start_time;
//
//    NSTimeInterval dateTime = startTime+timeOffset+[SystemTimeTable beginningOf1980];
//    NSDate* result = [[NSDate alloc] initWithTimeIntervalSince1970:dateTime];
//    result = [result dateAtNearestMinute];
//    return result;
//    }
//
//    +(NSDate*) endDateFromEventRecord:(EventInformationRecord*)eventRecord withTimeOffest:(NSTimeInterval)timeOffset
//    {
//    NSDate* start_time = [ScheduledShow startDateFromEventRecord:eventRecord withTimeOffest:timeOffset];
//
//    NSDate* result = [start_time dateByAddingTimeInterval:eventRecord.length_in_seconds];
//    return result;
//    }

    
    public func update(fromEventRecord eventRecord: EventInformationRecord, withTimeOffset timeOffset: TimeInterval)
    {
        guard let context = self.managedObjectContext else
        {
            return
        }
        
        context.performAndWait {
            let startTime = eventRecord.start_time
            let dateTime = startTime+timeOffset+SystemTimeTable.beginningOf1980()
            let start_date = Date(timeIntervalSince1970: dateTime).dateAtNearestMinute
            if let originalStartTime = self.start_time
            {
                if fabs(originalStartTime.timeIntervalSince(start_date)) > 120.0
                {
                        self.start_time = start_date as NSDate
                }
            }
            else
            {
                self.start_time = start_date as NSDate
            }
            
            let endTime = start_date.addingTimeInterval(eventRecord.length_in_seconds)
            self.end_time = endTime as NSDate
            self.event_id = NSNumber(value: eventRecord.event_id)
            let showTitleEntity = NSEntityDescription.entity(forEntityName: "ShowTitle", in: context)!
            for aLanguageString in eventRecord.workingTitles()
            {
                var foundIt = false
                if let titles = self.titles as? Set<LocalizedString>
                {
                    for aTitle in titles
                    {
                        if aTitle.locale == aLanguageString.languageCode
                        {
                            foundIt = true
                            aTitle.text = aLanguageString.string
                            break
                        }
                    }
                }
                if !foundIt
                {
                    let newTitle = ShowTitle(entity: showTitleEntity, insertInto: context)
                    newTitle.locale = aLanguageString.languageCode
                    newTitle.text = aLanguageString.string
                    newTitle.show = self
                    self.addToTitles(newTitle)
                }
                
                if self.contentAdvisories?.count == 0
                {
                    for aTable in eventRecord.descriptors
                    {
                        if let contentAdvisoryDescriptor = aTable as? ContentAdvisoryDescriptor
                        {
                            ContentAdvisory.extractAdvisories(fromDescriptors : contentAdvisoryDescriptor, intoShow: self)
                        }
                    }
                }
            }
            
        }
    }
    public func update(fromExtendedTextTable extendedText : ExtendedTextTable)
    {
        guard let context = self.managedObjectContext else
        {
            return
        }
        context.performAndWait {
            let showDescriptionEntity = NSEntityDescription.entity(forEntityName: "ShowDescription", in: context)!
            for aLanguageString in extendedText.strings
            {
                var foundIt = false
                if let descriptions = self.descriptions as? Set<LocalizedString>
                {
                    for aDescription in descriptions
                    {
                        if aDescription.locale == aLanguageString.languageCode && !aLanguageString.string.isEmpty
                        {
                            foundIt = true
                            aDescription.text = aLanguageString.string
                        }
                    }
                }
                if !foundIt
                {
                    let newDescription = ShowDescription(entity: showDescriptionEntity, insertInto: context)
                    
                    newDescription.locale = aLanguageString.languageCode
                    newDescription.text = aLanguageString.string
                    newDescription.show = self
                    self.addToDescriptions(newDescription)
                }
            }
        }
    }
}
